File: Parser/tokenizer.cc
#include "tokenizer.hh"

using namespace Fract::Parser;
using namespace Fract::Utilities;

tokenizer::tokenizer() { /**/ }

tokenizer::tokenizer(code_file *file) {
  finish = false;
  this->file = file;
  column = 1;
  line_iterator = file->lines.begin();
}

void
tokenizer::exit_tokenizer_error(std::string message) {
  std::cout << std::endl;
  std::cout << "TOKENIZER ERROR" << std::endl;
  std::cout << "MESSAGE: " << message << std::endl;
  std::cout << "LINE: " << line_iterator->line << std::endl;
  std::cout << "COLUMN: " << column << std::endl;
  exit(EXIT_FAILURE);
}

/* Last putted token */
token last_token;

token
tokenizer::next_token() {
  token _token;
  _token.line = line_iterator->line;
  _token.column = column;

  /* Return empty token is statement is finished. */
  if(column >= line_iterator->text.length()) {
    return _token;
  }

  std::string statement = line_iterator->text.substr(column - 1);

  /* Ignore whitespaces and tabs. */
  for(int index = 0; index < statement.length(); index++)
  { char ch = statement[index];
    if(
      ch == ' ' ||
      ch == '\t'
      ) {
      column++;
    }
    else {
      statement = statement.substr(index);
      break;
    }
  }

  /* Return empty token is statement is empty. */
  if(statement == "") {
    return _token;
  }

  /* Arithmetic value check */
  if(statement[0] == token_minus[0] || arithmetic::is_numeric(statement[0])) {
    std::string value = statement[0] == token_minus[0] ? token_minus : "";
    if(
      value == "" ||
      (value != "" && (
        last_token.type == type_operator          ||
        last_token.type == type_open_parenthes    ||
        last_token.type == type_close_parenthes
      ))
    )
    { int index = value.length();
      for(; index < statement.size(); index++)
      { char ch = statement[index];
        if(!arithmetic::is_numeric(ch) && ch != token_dot[0]) {
          break;
        }
        value += ch;
      }
      statement = value;
    }
  }

  /* Check anothers. */
  if(arithmetic::is_integer_number(statement))
  { _token.type = type_value;
    _token.value = statement;
  }
  else if(arithmetic::is_floating_number(statement))
  { _token.type = type_value;
    _token.value = statement;
  }
  else if(string::starts_with(statement, kw_variable))
  { _token.type = type_let;
    _token.value = kw_variable;
  }
  else if(string::starts_with(statement, token_plus))
  { _token.type = type_operator;
    _token.value = token_plus;
  }
  else if(string::starts_with(statement, token_minus))
  { _token.type = type_operator;
    _token.value = token_minus;
  }
  else if(string::starts_with(statement, token_star))
  { _token.type = type_operator;
    _token.value = token_star;
  }
  else if(string::starts_with(statement, token_slash))
  { _token.type = type_operator;
    _token.value = token_slash;
  }
  else if(string::starts_with(statement, token_lparenthes))
  { _token.type = type_open_parenthes;
    _token.value = token_lparenthes;
  }
  else if(string::starts_with(statement, token_rparenthes))
  { _token.type = type_close_parenthes;
    _token.value = token_rparenthes;
  }
  else {
    exit_tokenizer_error("What the?: '" + statement + "'");
  }

  column += _token.value.length();
  return _token;
}

std::vector<token>
tokenizer::tokenize_next() {
  std::vector<token> tokens;

  if(finish) {
    return tokens;
  }

  if(line_iterator->text == "") {
    return tokens;
  }

  /* Reset to defaults */
  column = 1;
  last_token.type = type_none;
  last_token.value = "";

  token _token;
  while((_token = next_token()).value != "") {
    tokens.push_back(_token);
    last_token = _token;
  }

  if(line_iterator == file->lines.end()) {
    finish = true;
  }

  line_iterator++;
  return tokens;
}

File: parser/tokenizer/tokenizer.go
package tokenizer

import (
	"fmt"
	"os"
	"strings"

	arithmeric "../../fract/arithmetic"
	"../../grammar"
	"../../objects"
)

// Tokenizer Tokenizer of Fract.
type Tokenizer struct {
	/* PRIVITE */

	// Source file.
	file *objects.CodeFile
	// Last column.
	column int
	// Index of last line.
	index int

	/* PUBLIC */

	// Finished destination file.
	Finish bool
}

// New Create new instance of Tokenizer.
// file Destination file.
func New(file *objects.CodeFile) *Tokenizer {
	var _tokenizer *Tokenizer = new(Tokenizer)
	_tokenizer.column = 1
	_tokenizer.file = file
	return _tokenizer
}

// ExitError Exit with tokenizer error.
// message Message of error.
func (t *Tokenizer) ExitError(message string) {
	fmt.Println()
	fmt.Println("TOKENIZER ERROR")
	fmt.Println("MESSAGE: " + message)
	fmt.Printf("LINE: %d", t.file.Lines[t.index].Line)
	fmt.Printf("COLUMN: %d", t.index)
	os.Exit(1)
}

/* Last putted token */
var lastToken objects.Token

// NextToken Tokenize next token from statement.
func (t *Tokenizer) NextToken() objects.Token {
	var _token objects.Token
	_token.Line = t.index + 1
	_token.Column = t.column

	var cline objects.CodeLine = t.file.Lines[t.index]

	/* Return empty token is statement finished. */
	if t.column >= len(cline.Text) {
		return _token
	}

	var statement string = cline.Text[t.column-1:]

	/* Ignore whitespaces and tabs */
	for index := 0; index < len(statement); index++ {
		var ch byte = statement[index]
		if ch == ' ' || ch == 't' {
			t.column++
		} else {
			statement = statement[index:]
			break
		}
	}

	/* Return empty token if statement is empty. */
	if statement == "" {
		return _token
	}

	/* Arithmetic value check. */
	if statement[0] == grammar.TokenMinus[0] || arithmeric.IsNumeric(statement[0]) {
		var value string = ""
		if statement[0] == grammar.TokenMinus[0] {
			value = grammar.TokenMinus
		}
		if value == "" || (
			value != "" && (
				lastToken.Type == TypeOperator      ||
				lastToken.Type == TypeOpenParenthes ||
				lastToken.Type == TypeCloseParenthes
			)
		) {
			for char := range statement[index:] {
				if !arithmeric.IsNumeric(char) && char != grammar.TokenDot[0] {
					break
				}
				value += char
			}
			stastatement = value
		}
	}

	/* Check anothers. */
	if arithmetic.IsInteger(statement) {
		_token.Type = TypeValue
		_token.Value = statement
	} else if arithmetic.IsFloat(statement) {
		_token.Type = TypeValue
		_token.Value = statement
	} else if strings.HasPrefix(statement, grammar.KwVariable) {
		_token.Type = TypeLet
		_token.Value = grammar.KwVariable
	} else if strings.HasPrefix(statement, grammar.TokenPlus) {
		_token.Type = TypeOperator
		_token.Value = grammar.TokenPlus
	} else if strings.HasPrefix(statement, grammar.TokenMinus) {
		_token.Type = TypeOperator
		_token.Value = grammar.TokenMinus
	} else if strings.HasPrefix(statement, grammar.TokenStar) {
		_token.Type = TypeOperator
		_token.Value = grammar.TokenStar
	} else if strings.HasPrefix(statement, grammar.TokenSlash) {
		_token.Type = TypeOperator
		_token.Value = grammar.TokenSlash
	} else if strings.HasPrefix(statement, grammar.TokenLParenthes) {
		_token.Type = TypeOpenParenthes
		_token.Value = grammar.TokenLParenthes
	} else if strings.HasPrefix(statement, grammar.TokenRParenthes) {
		_token.Type = TypeCloseParenthes
		_token.Value = grammar.TokenRParenthes
	} else {
		t.ExitError("What the?: '" + statement + "'")
	}

	t.column += len(_token.Value)
	return _token
}

// TokenizeNext Tokenize all statement.
func (t *Tokenizer) TokenizeNext() []objects.Token {
	var tokens []objects.Token

	if t.Finish {
		return tokens
	}

	if t.file.Lines[t.index].Text == "" {
		return tokens
	}

	/* Reset to defaults */
	t.column = 1
	lastToken.Type = TypeNone
	lastToken.Value = ""

	var _token objects.Token = t.NextToken()
	for _token.Value != "" {
		tokens = append(tokens, _token)
		lastToken = _token
		_token = t.NextToken()
	}

	if t.index == len(t.file.Lines)-1 {
		t.Finish = true
	}

	t.index++
	return tokens
}
